staging/src/k8s.io/legacy-cloud-providers/azure/clients/armclient/azure_armclient.go

```diff
@@ -0,0 +1,543 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package armclient
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"k8s.io/client-go/pkg/version"
+	"k8s.io/klog"
+	"k8s.io/legacy-cloud-providers/azure/retry"
+)
+
+var _ Interface = &Client{}
+
+// Client implements ARM client Interface.
+type Client struct {
+	client  autorest.Client
+	backoff *retry.Backoff
+
+	baseURI      string
+	apiVersion   string
+	clientRegion string
+}
+
+// New creates a ARM client
+func New(authorizer autorest.Authorizer, baseURI, userAgent, apiVersion, clientRegion string, clientBackoff *retry.Backoff) *Client {
+	restClient := autorest.NewClientWithUserAgent(userAgent)
+	restClient.PollingDelay = 5 * time.Second
+	restClient.RetryAttempts = 3
+	restClient.RetryDuration = time.Second * 1
+	restClient.Authorizer = authorizer
+
+	if userAgent == "" {
+		restClient.UserAgent = GetUserAgent(restClient)
+	}
+
+	backoff := clientBackoff
+	if backoff == nil {
+		// 1 steps means no retry.
+		backoff = &retry.Backoff{
+			Steps: 1,
+		}
+	}
+
+	return &Client{
+		client:       restClient,
+		baseURI:      baseURI,
+		backoff:      backoff,
+		apiVersion:   apiVersion,
+		clientRegion: NormalizeAzureRegion(clientRegion),
+	}
+}
+
+// GetUserAgent gets the autorest client with a user agent that
+// includes "kubernetes" and the full kubernetes git version string
+// example:
+// Azure-SDK-for-Go/7.0.1 arm-network/2016-09-01; kubernetes-cloudprovider/v1.17.0;
+func GetUserAgent(client autorest.Client) string {
+	k8sVersion := version.Get().GitVersion
+	return fmt.Sprintf("%s; kubernetes-cloudprovider/%s", client.UserAgent, k8sVersion)
+}
+
+// NormalizeAzureRegion returns a normalized Azure region with white spaces removed and converted to lower case
+func NormalizeAzureRegion(name string) string {
+	region := ""
+	for _, runeValue := range name {
+		if !unicode.IsSpace(runeValue) {
+			region += string(runeValue)
+		}
+	}
+	return strings.ToLower(region)
+}
+
+// sendRequest sends a http request to ARM service.
+// Although Azure SDK supports retries per https://github.com/azure/azure-sdk-for-go#request-retry-policy, we
+// disable it since we want to fully control the retry policies.
+func (c *Client) sendRequest(ctx context.Context, request *http.Request) (*http.Response, *retry.Error) {
+	sendBackoff := *c.backoff
+	response, err := autorest.SendWithSender(
+		c.client,
+		request,
+		retry.DoExponentialBackoffRetry(&sendBackoff),
+	)
+	return response, retry.GetError(response, err)
+}
+
+// Send sends a http request to ARM service with possible retry to regional ARM endpoint.
+func (c *Client) Send(ctx context.Context, request *http.Request) (*http.Response, *retry.Error) {
```

Comment:
it's a lot of code changes, did you copy from somewhere else? where is the source? is it possible to import that lib?

Reply:
I'm referring Azure Go SDK for most interfaces (e.g. the VMSS clients [here](https://github.com/Azure/azure-sdk-for-go/blob/master/services/compute/mgmt/2019-07-01/compute/virtualmachinescalesets.go), but we're unable to import them directly because we want to control the internal loops ourselves. 

See https://github.com/kubernetes-sigs/cloud-provider-azure/issues/247 for backgrounds why we need to do so.